### ISSUE TITLE
Marking datatype member function as "extern" results in a boogie error

### ISSUE THREAD

**Issue by danmatichuk** [dafny-lang/dafny#612]
The following Dafny program fails to translate:
```
datatype D = D
{
  function method {:extern Foo} Foo() : bool
}
```

Resulting in the following error:

```
Dafny 2.3.0.10506
(0,-1): Error: invalid type for argument 1 in application of _module.D.Foo: ref (expected: DatatypeType)
(0,-1): Error: invalid type for argument 1 in application of _module.D.Foo#requires: ref (expected: DatatypeType)
(0,-1): Error: invalid type for argument 1 in application of _module.D.Foo: ref (expected: DatatypeType)
Soundness.dfy(3,29): Error: invalid type for argument 1 in application of _module.D.Foo#Handle: DatatypeType (expected: ref)
4 type checking errors detected in /var/folders/zl/80pk530517747r2sb5_rcfm00000gn/T/Soundness__module.bpl

*** Encountered internal translation error - re-running Boogie to get better debug information

/var/folders/zl/80pk530517747r2sb5_rcfm00000gn/T/Soundness__module.bpl(2729,32): Error: invalid type for argument 1 in application of _module.D.Foo: ref (expected: DatatypeType)
/var/folders/zl/80pk530517747r2sb5_rcfm00000gn/T/Soundness__module.bpl(2734,38): Error: invalid type for argument 1 in application of _module.D.Foo#requires: ref (expected: DatatypeType)
/var/folders/zl/80pk530517747r2sb5_rcfm00000gn/T/Soundness__module.bpl(2741,23): Error: invalid type for argument 1 in application of _module.D.Foo: ref (expected: DatatypeType)
/var/folders/zl/80pk530517747r2sb5_rcfm00000gn/T/Soundness__module.bpl(2742,21): Error: invalid type for argument 1 in application of _module.D.Foo: ref (expected: DatatypeType)
/var/folders/zl/80pk530517747r2sb5_rcfm00000gn/T/Soundness__module.bpl(2745,50): Error: invalid type for argument 1 in application of _module.D.Foo#Handle: DatatypeType (expected: ref)
```

**Comment by davidcok** [comment]
Dan - the current master HEAD does not show this problem for me. Is it resolved for you?

**Comment by danmatichuk** [comment]
Yes. I just updated and it seems to have been resolved somewhere between c580b1e8 and 61813f20.

**Comment by davidcok** [comment]
Added the issue as a test case so it stays fixed.

**Comment by RustanLeino** [comment]
This was fixed as part of #651
